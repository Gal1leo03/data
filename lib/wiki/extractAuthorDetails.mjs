/**
 * Gets author details from a wikipedia page returned by the API
 */
export function extractAuthorDetails(page) {
  if (!page) return null
  const {
    extract,
    title,
    pageid,
    canonicalurl,
    pageimage,
    images,
    description: originalDescription = '',
    categories = [],
  } = page

  // The `extract` is a brief summary of the page, usually taken from the first
  // paragraph. We use this the for author's `bio` field.
  // NOTE: The length and content of the `extract` can be customized via API
  // parameters when requesting a page (see lib/wiki/api). Currently it is set
  // to two sentences.
  const bio = extract
    // clean up the extract...
    // In most cases, the extract starts with the person's full legal name,
    // followed by "(date of birth [- date of death])". Sometimes there is
    // additional text inside the parenthesis, such as alternate names,
    // pronunciations and spellings in different languages, etc, each separated
    // by a `;`. We try to remove everything from inside the parenthesis
    // except the birth and death dates. Example:
    //   Before: "( ; (listen); 15 October 1931 – 27 July 2015)"
    //   After:  "(15 October 1931 – 27 July 2015)"
    .replace(/\(\s*listen\s*\)/g, ' ')
    .replace(/\(([^;)]+;\s*)+/, '(')
    .replace(/\[.*\]/g, ' ')
    .replace(/\([;\s]+/, '(')
    // Remove newlines and double spaces from the extract
    .replace(/(\s+)|(\n)/g, ' ')

  // The `description` is a short, one line description of the person.
  // It's usually their primary occupation or what they are known for.
  // For example: "Theoretical Physicist"  or "American Author"
  // In some cases, the description also includes birth/death dates inside
  // parenthesis. We remove those if present.
  const description = originalDescription.replace(/\(.+\)/, '').trim() || null

  // Get a list of category names that the person belongs to. In the future,
  // we could use this to add tags/categories to authors.
  const categoryNames = categories.map(category => {
    return category.title.replace(/category:/i, '').trim()
  })

  // This only includes fields we need for Authors.
  return {
    name: title,
    pageId: pageid,
    link: canonicalurl,
    bio,
    description,
    imagePath: pageimage,
    categories: categoryNames,
    images,
  }
}
