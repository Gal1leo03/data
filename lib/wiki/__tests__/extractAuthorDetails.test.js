import { extractAuthorDetails } from '../extractAuthorDetails.mjs'
import examplePage from '../__fixtures__/example-page.json'

describe('wiki.extractAuthorDetails', () => {
  it(`Returns the expected response`, async () => {
    const authorDetails = extractAuthorDetails(examplePage)
    expect(authorDetails).toMatchSnapshot()
  })
})
