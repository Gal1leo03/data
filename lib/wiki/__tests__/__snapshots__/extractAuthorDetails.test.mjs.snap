// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wiki.extractAuthorDetails Returns the expected response 1`] = `
Object {
  "bio": "Avul Pakir Jainulabdeen Abdul Kalam (15 October 1931 – 27 July 2015) was an Indian aerospace scientist who served as the 11th President of India from 2002 to 2007. He was born and raised in Rameswaram, Tamil Nadu and studied physics and aerospace engineering. He spent the next four decades as a scientist and science administrator, mainly at the Defence Research and Development Organisation (DRDO) and Indian Space Research Organisation (ISRO) and was intimately involved in India's civilian space programme and military missile development efforts.",
  "categories": Array [
    "1931 births",
    "2015 deaths",
    "20th-century Indian engineers",
    "20th-century Indian politicians",
    "21st-century Indian engineers",
    "21st-century Indian politicians",
    "A. P. J. Abdul Kalam",
    "All Wikipedia articles written in Indian English",
    "Articles with Encyclopædia Britannica links",
    "Articles with hAudio microformats",
    "Articles with hCards",
    "Articles with short description",
    "CS1 maint: archived copy as title",
    "CS1 maint: extra text: authors list",
    "CS1 maint: others",
    "Defence Research and Development Organisation",
    "Engineers from Tamil Nadu",
    "Fellows of the National Academy of Medical Sciences",
    "Good articles",
    "Indian Muslims",
    "Indian Space Research Organisation people",
    "Indian Tamil academics",
    "Indian Tamil politicians",
    "Indian aerospace engineers",
    "Madras Institute of Technology alumni",
    "Nuclear energy in India",
    "Pages including recorded pronunciations",
    "Pages using Sister project links with hidden wikidata",
    "People from Ramanathapuram district",
    "Presidents of India",
    "Recipients of the Bharat Ratna",
    "Recipients of the Padma Bhushan in civil service",
    "Recipients of the Padma Vibhushan in science & engineering",
    "Short description is different from Wikidata",
    "St Joseph's College, Tiruchirappalli alumni",
    "State funerals in India",
    "Tamil engineers",
    "Tamil poets",
    "University of Madras alumni",
    "Use Indian English from June 2016",
    "Use dmy dates from March 2021",
    "Webarchive template wayback links",
    "Wikipedia articles with BNF identifiers",
    "Wikipedia articles with CANTIC identifiers",
    "Wikipedia articles with GND identifiers",
    "Wikipedia articles with ISNI identifiers",
    "Wikipedia articles with LCCN identifiers",
    "Wikipedia articles with NDL identifiers",
    "Wikipedia articles with NKC identifiers",
    "Wikipedia articles with NLA identifiers",
    "Wikipedia articles with NLI identifiers",
    "Wikipedia articles with NLK identifiers",
    "Wikipedia articles with NTA identifiers",
    "Wikipedia articles with SUDOC identifiers",
    "Wikipedia articles with Trove identifiers",
    "Wikipedia articles with VIAF identifiers",
    "Wikipedia articles with WORLDCATID identifiers",
    "Wikipedia indefinitely semi-protected pages",
  ],
  "description": "Scientist and 11th President of India",
  "imagePath": "A._P._J._Abdul_Kalam.jpg",
  "images": undefined,
  "link": "https://en.wikipedia.org/wiki/A._P._J._Abdul_Kalam",
  "name": "A. P. J. Abdul Kalam",
  "pageId": 62682,
}
`;
